phase2/outline
# Feature: Bubble events

## Description

Add a way to bubble / redispatch events up the DOM tree so that things like Google Tag Manager can watch for them.

For now, included `submit` as a default option.

The new custom events provide
- The same `event.type` name
- An attempt to provide getters for all the properties on the source event
- `event.sourceEvent` which is the original event and can take functions like `preventDefault()`
- `event.aggregatedComposedPath()` which is a substitute for `event.composedPath()` that attempts to show the full path through the DOM

## Type of change


## How Has This Been Tested?

- Visit http://localhost:6042/?path=/story/tests-bubble-events-submit--submit
- Enable `Preserve log` in Chrome inspector (see below)
- Click on the submit button
- View the logged event

![Enable_preserve_log](https://user-images.githubusercontent.com/397902/157557400-859fbc52-fe14-405e-b8f0-cb27ca7188dc.jpg)

![story](https://user-images.githubusercontent.com/397902/158849506-02d789bc-408b-4bc6-a6cf-75e45ce9b148.jpg)

![log_messages](https://user-images.githubusercontent.com/397902/158847151-ae3f9b65-e87c-4603-b7c6-b8c73652261e.jpg)

## Checklist:


## Notes

### Testing
I'm not sure if there is a better way to provide a test for this, so happy to hear other suggestions.

I know some projects (including the one that needs something like this) have an `x-form` component and if we get that into Outline, the test story / component could be deleted.

One of the key parts of testing was testing nested shadow DOMs (not slots). Wanted to make sure that the events can reach beyond a single shadow DOM and that we also only get one event per context. The provided test allowed me to check those conditions.

### Context

This is done as work for a client project, but it will hopefully resolve an issue for all projects that use Outline and we can easily define additional event types as needed.

<a href="https://gitpod.io/#https://github.com/phase2/outline/pull/262"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

### Attempted alternative ways to clone events

I tried `Proxy` objects, but I couldn't find a way to dispatch those.

I tried to clone the object with `new event.constructor(event.type, event);`, but I could not set the `bubbles` or `composed` properties which are necessary so that the event continues to bubble through multiple shadow DOMs.

I tried `Object.defineProperty(eventForLightDOM, 'composed', true)`, but that didn't seem to actually make the event bubble.

There may be a better way. Please let me know or try it out with the provided test story.
